File: bootstrap/token.ml
type token =
  | Atom of string
  | Comma
  | EOF
  | Eq
  | Error of string
  | FatArrow
  | Ident of string
  | Int of int
  | LParen of bool
  | Minus
  | Plus
  | RParen
  | Semi
  | Str of string

let token_to_string t =
  match t with
  | Atom s -> Printf.sprintf ":%s" s
  | Comma -> ","
  | EOF -> "EOF"
  | Eq -> "="
  | Error s -> Printf.sprintf "Error(%s)" s
  | FatArrow -> "=>"
  | Ident i -> i
  | Int i -> Printf.sprintf "%d" i
  | LParen imm ->
      if imm then
        "(i"
      else
        "("
  | Minus -> "-"
  | Plus -> "+"
  | RParen -> ")"
  | Semi -> ";"
  | Str s -> Printf.sprintf "\"%s\"" s

let tokens_to_string list = String.concat " " (List.map token_to_string list)

let tokenize s =
  let len = String.length s in
  let rec step i prev_ws =
    if i >= len then
      (EOF, i)
    else
      match String.get s i with
      | '+' -> (Plus, i + 1)
      | '-' -> (Minus, i + 1)
      | ';' -> (Semi, i + 1)
      | ',' -> (Comma, i + 1)
      | ':' ->
          let rec loop j =
            if j >= len then
              if j == i + 1 then
                (Error "empty atom", j)
              else
                (Atom (String.sub s (i + 1) (j - i - 1)), j)
            else
              match String.get s j with
              | 'a' .. 'z' | '_' | '0' .. '9' -> loop (j + 1)
              | _ -> (Atom (String.sub s (i + 1) (j - i - 1)), j)
          in
          loop (i + 1)
      | '=' ->
          if i + 1 < len && String.get s (i + 1) = '>' then
            (FatArrow, i + 2)
          else
            (Eq, i + 1)
      | '(' -> (LParen (prev_ws = false), i + 1)
      | ')' -> (RParen, i + 1)
      | '0' .. '9' ->
          let rec loop j =
            if j >= len then
              (Int (int_of_string (String.sub s i (j - i))), j)
            else
              match String.get s j with
              | '0' .. '9' -> loop (j + 1)
              | _ -> (Int (int_of_string (String.sub s i (j - i))), j)
          in
          loop (i + 1)
      | '"' ->
          let rec loop j =
            if j >= len then
              (Error "unterminated string literal", j)
            else
              match String.get s j with
              | '"' -> (Str (String.sub s (i + 1) (j - i - 1)), j + 1)
              | '\n' -> (Error "unterminated string literal", j)
              | _ -> loop (j + 1)
          in
          loop (i + 1)
      | '\n' | '\r' | ' ' -> step (i + 1) true
      | 'a' .. 'z' ->
          let rec loop j =
            if j >= len then
              (Ident (String.sub s i (j - i)), j + 1)
            else
              match String.get s j with
              | 'a' .. 'z' | '_' | '0' .. '9' | 'A' .. 'Z' -> loop (j + 1)
              | _ -> (Ident (String.sub s i (j - i)), j)
          in
          loop (i + 1)
      | c -> (Error (Printf.sprintf "unknown token: %c" c), i + 1)
  in
  let rec collect i acc =
    let t, i = step i false in
    match t with
    | EOF -> List.rev acc
    | _ -> collect i (t :: acc)
  in
  collect 0 []

File: bootstrap/token.py
from dataclasses import dataclass
from enum import Enum

from . import error
from .span import Span

Value = str | None


class Kind(Enum):
    behaviour_ns = "behaviour ns"
    braket_left = "["
    braket_right = "]"
    case = "case"
    char_lit = "char literal"
    colon = ":"
    comma = ","
    comment = "--"
    curly_left = "{"
    curly_right = "}"
    dot = "."
    else_ = "else"
    end = "end"
    eof = "end of file"
    eq = "="
    eqeq = "=="
    false = "false"
    fun = "fun"
    gt = ">"
    ident = "identifier"
    if_ = "if"
    int_lit = "int literal"
    lt = "<"
    minus = "-"
    neq = "!="
    paren_left = "("
    paren_right = ")"
    pipe = "|"
    plus = "+"
    str_lit = "str literal"
    true = "true"
    type_ident = "type identifier"


@dataclass
class Token:
    kind: Kind
    span: Span
    value: Value

    def __str__(self) -> str:
        value = ""
        match self.value:
            case str():
                value = self.value
            case None:
                value = ""
            case _:
                raise AssertionError(f"Unexpected value: {self.value}")
        return f"{self.span}: [{self.kind.value}] {value}"

    def value_str(self) -> str:
        assert self.value is not None, "Token has no value"
        return self.value


keywords = {
    x.value: x
    for x in (
        Kind.case,
        Kind.else_,
        Kind.end,
        Kind.false,
        Kind.fun,
        Kind.if_,
        Kind.true,
    )
}


@dataclass
class Input:
    file: str
    src: str
    index: int = 0

    def next(self) -> str:
        if self.index > len(self.src):
            return ""
        c = self.src[self.index]
        self.index += 1
        return c

    def peek(self) -> str:
        return "" if self.index == len(self.src) else self.src[self.index]

    def has_next(self) -> bool:
        return self.index < len(self.src)

    def span(self) -> Span:
        return Span(self.file, self.src, self.index, self.index)


def tokenize(input: Input) -> tuple[list[Token], list[error.Error]]:
    tokens: list[Token] = []
    errors: list[error.Error] = []
    while input.has_next():
        span = input.span()
        c = input.next()
        kind: Kind
        value: Value = None
        match c:
            case "":
                kind = Kind.eof
            case "\n" | "\r" | "\t" | " ":
                continue
            case "(":
                kind = Kind.paren_left
            case ")":
                kind = Kind.paren_right
            case "{":
                kind = Kind.curly_left
            case "}":
                kind = Kind.curly_right
            case "[":
                kind = Kind.braket_left
            case "]":
                kind = Kind.braket_right
            case ".":
                kind = Kind.dot
            case ",":
                kind = Kind.comma
            case "+":
                kind = Kind.plus
            case "<":
                kind = Kind.lt
            case ">":
                kind = Kind.gt
            case "|":
                kind = Kind.pipe
            case ":":
                kind = Kind.colon
            case "!":
                if input.peek() == "=":
                    input.next()
                    kind = Kind.neq
                else:
                    errors.append(error.unknown_token(span, c))
                    continue
            case "=":
                match input.peek():
                    case "=":
                        input.next()
                        kind = Kind.eqeq
                    case _:
                        kind = Kind.eq
            case "-":
                if input.peek() == "-":
                    # Comment
                    input.next()
                    kind = Kind.comment
                    value = "--"
                    while (c := input.peek()) not in ("", "\n"):
                        input.next()
                        value += c
                elif input.peek().isnumeric():
                    # Negative number
                    value = c
                    while (c := input.peek()).isnumeric():
                        input.next()
                        value += c
                    kind = Kind.int_lit
                else:
                    kind = Kind.minus
            case '"':
                # String
                # todo: support utf-8 characters
                # todo: support \ escapes
                value = ""
                while (c := input.peek()) not in ("", '"', "\n"):
                    input.next()
                    value += c
                if (c := input.peek()) in ("\n", ""):
                    errors.append(error.unterminated_str_lit(span.merge(input.span()), eof=c == ""))
                    continue
                input.next()
                kind = Kind.str_lit
            case "'":
                # Char
                # todo: support utf-8 characters
                # todo: support \ escapes
                value = input.next()
                if (c := input.peek()) != "'":
                    errors.append(error.unterminated_char_lit(span.merge(input.span()), eof=c == ""))
                    continue
                input.next()
                kind = Kind.char_lit
            case "@":
                kind = Kind.behaviour_ns
                value = ""
                while ((c := input.peek()).isalnum() and c.isascii()) or c == "_":
                    input.next()
                    value += c
            case c if c.isnumeric():
                # Int
                value = c
                while (c := input.peek()).isnumeric():
                    input.next()
                    value += c
                kind = Kind.int_lit
            case c if c.isalpha():
                # Identifier or type identifier
                kind = Kind.ident if c.islower() else Kind.type_ident
                value = c
                while ((c := input.peek()).isalnum() and c.isascii()) or c == "_":
                    input.next()
                    value += c
                keyword = keywords.get(value)
                if keyword:
                    kind = keyword
                    value = None
            case _:
                errors.append(error.unknown_token(span, c))
                continue
        tokens.append(Token(kind, span.merge(input.span()), value))
    tokens.append(Token(Kind.eof, input.span(), None))
    return tokens, errors
